File: 173.Binary_Search_Tree_Iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class BSTIterator {
    private Stack<TreeNode> stack = new Stack();
    
    public BSTIterator(TreeNode root) {
        pushAllLeft(root);
    }
    
    /** @return the next smallest number */
    public int next() {
        TreeNode curr = stack.pop();
        pushAllLeft(curr.right);
        return curr.val;
        
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        if (stack.isEmpty()) return false;
        return true;
    }
    
    private void pushAllLeft(TreeNode root) {
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 173.BinarySearchTreeIterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self.appendNode(root)

    def next(self) -> int:
        node = self.stack.pop()
        self.appendNode(node.right)
        return node.val

    def hasNext(self) -> bool:
        return len(self.stack) != 0

    def appendNode(self, root):
        while root:
            self.stack.append(root)
            root = root.left


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
